otter/util/zk.py

```diff
@@ -0,0 +1,115 @@
+from functools import partial
+
+from characteristic import attributes
+
+from effect import TypeDispatcher
+from effect.twisted import deferred_performer
+
+from kazoo.exceptions import NoNodeError, NodeExistsError
+
+from twisted.internet.defer import gatherResults
+
+
+@attributes(['path', 'content'])
+class CreateOrSet(object):
+    """
+    Create a node, or if the node already exists, set the content.
+
+    Handles the case where a node gets deleted in between our attempt and
+    creating and setting.
+    """
+
+
+@deferred_performer
+def perform_create_or_set(kz_client, dispatcher, create_or_set):
+    """
+    Performer for :obj:`CreateOrSet`. Must be partialed with ``kz_client``.
+    """
+    path = create_or_set.path
+    content = create_or_set.content
+
+    def create():
+        d = kz_client.create(path, content, makepath=True)
+        d.addErrback(_handle(NodeExistsError, set_content))
+        return d
+
+    def set_content():
+        d = kz_client.set(path, content)
+        d.addErrback(_handle(NoNodeError, create))
```

Comment:
Non-blocking: This can go in infinite loop. I do realize that it shouldn't if the exceptions returned are accurate. But I am not sure if you want to allow this path to exist.